Lib/pty.py

```diff
@@ -125,14 +147,90 @@ def _read(fd):
     """Default read function."""
     return os.read(fd, 1024)
 
-def _copy(master_fd, master_read=_read, stdin_read=_read):
-    """Parent copy loop.
+def _getmask():
+    """Get signal mask of current thread."""
+    return signal.pthread_sigmask(signal.SIG_BLOCK, [])
+
+def _sigblock():
+    """Block all signals."""
+    signal.pthread_sigmask(signal.SIG_BLOCK, ALL_SIGNALS)
+
+def _sigreset(saved_mask):
+    """Restore signal mask."""
+    signal.pthread_sigmask(signal.SIG_SETMASK, saved_mask)
+
+def _setup_pty(slave_echo):
+    """Open and setup a pty pair.
+
+    If current stdin is a tty, then apply current stdin's
+    termios and winsize to the slave and set current stdin to
+    raw mode. Return (master, slave, original stdin mode/None,
+    stdin winsize/None)."""
+    stdin_mode = None
+    stdin_winsz = None
+    try:
+        mode = tty.tcgetattr(STDIN_FILENO)
+    except tty.error:
+        stdin_tty = False
+        fd = slave_fd
+
+        master_fd, slave_fd, slave_path = openpty(name=True)
+        mode = tty.tcgetattr(slave_fd)
+    else:
+        stdin_tty = True
+        fd = STDIN_FILENO
+
+        stdin_mode = list(mode)
+        if HAVE_WINSZ:
+            stdin_winsz = tty.tcgetwinsize(STDIN_FILENO)
+
+    if slave_echo:
+        mode[tty.LFLAG] |= tty.ECHO
+    else:
+        mode[tty.LFLAG] &= ~tty.ECHO
+
+    if stdin_tty:
+        master_fd, slave_fd, slave_path = openpty(mode, winsz, True)
+        tty.cfmakeraw(mode)
+
+    tty.tcsetattr(fd, tty.TCSAFLUSH, mode)
+
+    return master_fd, slave_fd, slave_path, stdin_mode, stdin_winsz
+
+def _setup_winch(slave_path, saved_mask, handle_winch):
+    """Install SIGWINCH handler.
+
+    Returns old SIGWINCH handler if relevant; returns
+    None otherwise."""
+    old_hwinch = None
+    if handle_winch:
+        def hwinch(signum, frame):
+            """Handle SIGWINCH."""
+            _sigblock()
+            new_slave_fd = os.open(slave_path, os.O_RDWR)
+            tty.setwinsize(new_slave_fd, tty.getwinsize(STDIN_FILENO))
+            os.close(new_slave_fd)
+            _sigreset(saved_mask)
+
+        try:
+            # Raises ValueError if not called from main thread.
+            old_hwinch = signal.signal(signal.SIGWINCH, hwinch)
+        except ValueError:
+            pass
+
+    return old_hwinch
+
+def _copy(master_fd, master_read=_read, stdin_read=_read, \
+          saved_mask=set()):
+    """Parent copy loop for pty.spawn().
     Copies
             pty master -> standard output   (master_read)
             standard input -> pty master    (stdin_read)"""
     fds = [master_fd, STDIN_FILENO]
-    while fds:
-        rfds, _wfds, _xfds = select(fds, [], [])
+    while True:
```

Comment:
I do not think that `fds` is ever empty in this loop. Therefore, I replaced `while fds:` with the original `while True:`. Please correct me if I am wrong.